src/azul/indexer/index_controller.py

```diff
@@ -133,6 +133,9 @@ def _validate_notification(self, notification):
     def contribute(self, event, retry=False):
         for record in event:
             message = json.loads(record.body)
+            if message.get('attempts', 0) > 0 and not retry:
+                log.warning(f"Non-retry worker skipping message {message}.")
+                raise Exception('Non-retry worker rejects previously attempted messages.')
             attempts = record.to_dict()['attributes']['ApproximateReceiveCount']
```

Comment:
This should be the sum of the `attempts` message property and the approximate receive count. It should reflect the number of actual attempts to process the message. So the "skipping" by the redrive lambda should not count against it. But the externally managed retries of messages in the retry queue, should. In some cases the approximate receive count can only be 0 (or 1, I forget) because the driving queue has a max receive count of 1. You should assert that as well.

Reply:
@hannes-ucsc: "Always have the attempts property in the message, initializing it to one"
```
sequence: `normal` -> `retry` -> `redrive` -> `fail`
x is `ApproximateReceiveCount`
y is `message.attempts`
z is # of attempts for logging

z = (1 if x > 1 else 0) + y
x, y -> z
1, 1 -> 1
2, 1 -> 2
3, 1 -> 2

1, 2 -> 2
2, 2 -> 3
3, 2 -> 3

1, 3 -> 3
2, 3 -> 4
3, 3 -> 4

1, 4 -> 4
2, 4 -> 5
3, 4 -> 5

1, 5 -> 5
2, 5 -> 6
3, 5 -> 6

1, 6 -> 6
2, 6 -> 7
3, 6 -> 7

1, 7 -> 7
2, 7 -> 8
3, 7 -> 8

1, 8 -> 8
2, 8 -> 9
3, 8 -> 9

1, 9 -> 9
2, 9 -> 10
3, 9 -> 10
```

The value of `z` should only be regarded for logging purposes. Decisions should be based on `y` only.